### PR TITLE
Clear sky 3D radiative heat tendencies added as output variables with CAM and all RRTMG radiation schemes

### PR DESCRIPTION
**Pull request by jhruppert**
TYPE: enhancement

KEYWORDS: radiation, clear-sky heating, CAM, RRTM, RRTMG, RRTMG-fast, RRTMG-K

SOURCE: James Ruppert (Penn State University)

DESCRIPTION OF CHANGES: 
Three-dimensional clear sky radiative heat tendencies are made available as output variables when using the following 
radiation schemes: 
   * RRTM longwave (ra_lw_phys=1)
   * CAM shortwave and longwave (ra_sw_phys, ra_lw_phys=3)
   * RRTMG shortwave and longwave (ra_sw_phys, ra_lw_phys=4)
   * RRTMG-fast version shortwave and longwave (ra_sw_phys, ra_lw_phys=24)
   * RRTMG-K shortwave and longwave (ra_sw_phys, ra_lw_phys=14). 

Two new variables are added only to the restart stream in Registry.EM_COMMON, defined as 
"UNCOUPLED THETA TEND DUE TO CLEAR SKY LONG [or SHORT] WAVE RAD", and with the units "K s-1". 
   * RTHRATENLWC ("RTHRATLWC" in history files)
   * RTHRATENSWC ("RTHRATSWC")

In the case of the three RRTMG schemes, the clear sky tendencies were already calculated, and hence just needed to 
be passed up to the main radiation driver (module_radiation_driver.F) and RK driver (module_first_rk_step_part1.F). 

For CAM, the clear-sky fluxes were calculated but not the corresponding heating rates. The appropriate code was added 
to do so, mimicking the calculations for all-sky but invoking only clear-sky fluxes.

LIST OF MODIFIED FILES:
M       Registry/Registry.EM_COMMON
M       dyn_em/module_first_rk_step_part1.F
M       dyn_nmm/module_PHYSICS_CALLS.F
M       phys/module_ra_cam.F
M       phys/module_ra_rrtm.F
M       phys/module_ra_rrtmg_lw.F
M       phys/module_ra_rrtmg_lwf.F
M       phys/module_ra_rrtmg_sw.F
M       phys/module_ra_rrtmg_swf.F
M       phys/module_ra_rrtmg_swk.F
M       phys/module_radiation_driver.F

TESTS CONDUCTED: 
1. Compiled okay with Intel. Verified that the clear-sky radiative heat tendencies are nearly identical to corresponding all-sky tendencies on the large scale, with much less variance at the cloud scales. All-sky total radiative heat tendency (RTHRATEN) is unchanged, as are the uncoupled longwave and shortwave tendencies (RTHRATENLW, RTHRATENSW).
2. Jenkins testing OK

RELEASE NOTE: Three-dimensional clear sky radiative heat tendencies are now available as output variables when using the following radiation schemes: RRTM longwave (ra_lw_phys=1), CAM shortwave and longwave (ra_sw_phys, ra_lw_phys=3), RRTMG shortwave and longwave (ra_sw_phys, ra_lw_phys=4), RRTMG-fast version shortwave and longwave (ra_sw_phys, ra_lw_phys=24), and RRTMG-K shortwave and longwave (ra_sw_phys, ra_lw_phys=14). Similar to the all-sky longwave and shortwave radiative heat tendencies (RTHRATLW and RTHRATSW), the clear-sky tendencies are not included in the default history output stream, but can now be added to it. They can be found in Registry.EM_COMMON as RTHRATLWC for longwave clear-sky heating rate and RTHRATSWC for shortwave clear-sky heating rate.

### CODE REVIEW

**Comment by jhruppert** [comment]
I also incorporated bug fix #1257 in this latest commit.

**Comment by davegill** [comment]
@jhruppert 
We do an automatic test with pull requests. Here is the output:
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 6683f7813a4db7ff0f8819edee8820bec5dddefb, requested by: jhruppert for PR: https://github.com/wrf-model/WRF/pull/1258. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 166           164         3
    Number of Comparisons  : 105           104         0

    Failed Simulations are: 
    output_2:2 = STATUS test_002m hwrf nmm_real 34 1NE
output_2:2 = STATUS test_002m hwrf nmm_real 34 2NE
output_2:2 = STATUS test_002m hwrf nmm_real 34 3NE
    Which comparisons are not bit-for-bit: 
    None
```
This indicates that there is a failure for the NMM dycore build/run (usually the "build" is the issue).

It is best to see what the problem is. Here are some simple instructions to build the NMM code:
https://github.com/davegill/wrf-coop/blob/master/README_user.md#WRFNMM

Almost always the trouble with NMM is when arguments have been added to a scheme from the physics driver. There are two different routines (one for ARW and one for NMM) that need to be considered. Since you modified dyn_em/module_rk_step_part1.F, you probably need to also modify dyn_nmm/module_PHYSICS_CALLS.F. Similarly, you modified the EM_COMMON regiistry, we probably need the NMM registry changed, too.

Anyways - first find out what is failing, and then we can make some suggestions.

**Comment by weiwangncar** [comment]
@jhruppert There is no need to add the radiative tendency fix into this PR.

**Comment by jhruppert** [comment]
@davegill Thanks for pointing me to dyn_nmm/module_PHYSICS_CALLS.F. That was indeed the part that was breaking. I've now added local variables inside the NMM physics driver to accommodate the changes in the call to phys/module_radiation_driver.F, and it now compiles successfully on my machine.

**Comment by dudhia** [comment]
Also check if they can be made optional instead of changing NMM's call.

File: dyn_nmm/module_PHYSICS_CALLS.F [hunk 1]
```diff
@@ -197,6 +197,8 @@ SUBROUTINE RADIATION(NTSD,DT,JULDAY,JULYR,XTIME,JULIAN            &
                                                         SWNIRDIR, &
                                                         SWNIRDIF
 
+      REAL,DIMENSION(IMS:IME,KMS:KME,JMS:JME) :: RTHRATENLWC
```
Comment (davegill) [inline dyn_nmm/module_PHYSICS_CALLS.F:200]: @jhruppert 
Are you able to add in the INTENT for these clear sky variables?

**Comment by smileMchen** [comment]
@davegill @dudhia @jhruppert 
I wonder whether someone is running regression test? I have received all the emails from the classroom.

**Comment by davegill** [comment]
Status of jenkins is all pass:
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 5570de06b73f7b6ed964c10aacd7d2ac9421617f, requested by: jhruppert for PR: https://github.com/wrf-model/WRF/pull/1258. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 166           164        0
    Number of Comparisons  : 105           104        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

**Comment by davegill** [comment]
@smileMchen 
> I wonder whether someone is running regression test? I have received all the emails from the classroom.

Ming,
It was me! I wondered where the email went. I must have forgotten to change the email address. Sorry. It looks like they all passed.

**Comment by smileMchen** [comment]
@davegill 
Dave, thanks for letting me know. Yes they all pass. I will delete the emails. 
Ming

**Comment by davegill** [comment]
Jenkins status is all PASS
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 041a43c315d0be51ba1e3ccc48824e6d0c42fd10, requested by: jhruppert for PR: https://github.com/wrf-model/WRF/pull/1258. For any query please send e-mail to David Gill.


    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 166           164        0
    Number of Comparisons  : 105           104        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

File: dyn_nmm/module_PHYSICS_CALLS.F (hunk 1 above)
Comment (jhruppert) [reply dyn_nmm/module_PHYSICS_CALLS.F]: @davegill 
Dave, no problem, just added these.

**Comment by davegill** [comment]
@jhruppert 
The "intent" mods broke the NMM, probably the build. Take a look at:
https://github.com/davegill/wrf-coop/blob/master/README_user.md#WRFNMM
These are the few rules to build the WRF model with the NMM dy-core.

**Comment by davegill** [comment]
@jhruppert 
The file that seems to be missing from the ordered grouping is LW for "k". We have sw, swf, swk. We have lw, lwf. No lwk file is modified. Just verifying that this was intentional and not an oversight.

**Comment by jhruppert** [comment]
@davegill 
I just checked back to confirm this, and that is correct. The reason there's no modified file module_ra_rrtmg_lwk.F is that the main driver for the RRTMG-K scheme (rad_rrtmg_driver) is inside module_ra_rrtmg_swk.F. This driver was already receiving the variable hrc (clear sky LW tendency) from the LW subroutines, so I just needed to get that passed up to the main radiation driver.

**Comment by weiwangncar** [comment]
This one looks ok to me.

**Comment by davegill** [comment]
@weiwangncar @dudhia 
These new fields are entirely diagnostic, right? These have been added to the restart stream, not history. So, after this PR, a user still has no access to these fields. The Registry I/O switch "r" should be "h", correct?

**Comment by weiwangncar** [comment]
@davegill Like other tendency fields, they are not available in standard history file. If a user wants them, they have to either modify the registry or use runtime io to get them. What he did here is consistent with what we have for other tendency fields.

**Comment by davegill** [comment]
@weiwangncar 
> What he did here is consistent with what we have for other tendency fields.

Wei, 
OK, good for me.

**Review by weiwangncar — APPROVED** [review]

**Comment by davegill** [comment]
jenkins is OK
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 5c16444d18026a4e2508e89af017ba3f1904f4d6, requested by: jhruppert for PR: https://github.com/wrf-model/WRF/pull/1258. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 163           161        0
    Number of Comparisons  : 103           102        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

**Pull request merged**